### PR TITLE
Sorting, Limiting, Filtering, and Paging for GET APIs

### PR DESCRIPTION
**Pull request by dwasinge**
# Changes to Existing Endpoints

`GET /engagements` 

Now supports the following parameters:

- Header Params
    - `Accept-version` - used to specify default results per page

- Query Params
    - `include` - attributes to include in response
    - `exclude` - attributes to exclude in response
    - `page` - page number to retrieve
        - if provided, the specified page will be returned.  defaults to 1
    - `perPage` - number of records to retrieve for each page
        - if provided, the specified number of records will be returned
        - if header `Accept-version` is missing or set to `v1`, defaults to 500
        - if header `Accept-version` is specified and not `v1`, defaults to 20
    - `search` - query string to filter engagements
        - supports `=`, `like`, `not like`, `exists`, and `not exists`
        - `start` and/or `end` can be used to limit the results based on a date range (i.e. `start=2021-01-01&end=2021-05-01`)
    - `sortOrder` - ASC for ascending and DESC for descending.  defaults to descending
    - `sortFields` - fields to sort on.  defaults to `customer_name,project_name`

`GET /engagements/customers/suggest`
`GET /engagements/artifacts/types`
`GET /engagements/categories`

Now supports the following parameters:

- Query Params
    - `page` - page number to retrieve
        - if provided, the specified page will be returned.  defaults to 1
    - `perPage` - number of records to retrieve for each page
        - if provided, the specified number of records will be returned
        - if header `Accept-version` is missing or set to `v1`, defaults to 500
        - if header `Accept-version` is specified and not `v1`, defaults to 20
    - `suggestion` - case insensitive query string to filter engagements
    - `sortOrder` - ASC for ascending and DESC for descending.  defaults to descending

# New Endpoints:

`GET /engagements/state/{state}` - returns engagements in the provided state

- Path Params 
    - `state` - the state to filter engagements on
        - values are `upcoming`, `active`, `past`, and `terminating`
        - if unknown value supplied, `upcoming` will be used

- Header Params
    - supports all the same as `GET /engagements`

- Query Params
    - supports all the same as `GET /engagements`

`GET /engagements/users/summary` - returns a summary count that includes the number of all users, the number of red hat users, and the number of other users

- Query Params:
    - `search` - query string to filter engagements
        - supports `=`, `like`, `not like`, `exists`, and `not exists`
        - `start` and/or `end` can be used to limit the results based on a date range (i.e. `start=2021-01-01&end=2021-05-01`)

### CODE REVIEW

**Review by mcanoy — COMMENTED** [review]
How performant is this? Will it scale ok?

File: src/main/java/com/redhat/labs/lodestar/model/pagination/PagedEngagementResults.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package com.redhat.labs.lodestar.model.pagination;
```
[the remark this answers is not in the document]
Comment (dwasinge) [reply src/main/java/com/redhat/labs/lodestar/model/pagination/PagedEngagementResults.java:1]: I originally tried to do that, but was having issues with using the Generic Type  in the method: `mongoCollection().aggregate(pipeline, Results.class)`.  There might be a way to cast it appropriately, but I didn't go back once it was working.  There will be more of these classes required in the next PR for the other nested objects so I will try to find a better solution.

File: src/main/java/com/redhat/labs/lodestar/model/search/StateSearchComponent.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    /**
+     * Returns a {@link Bson} for the configured state.
+     * 
+     * @return
+     */
+    private Bson createSearchBson() {
+
+        Bson launched = null;
+        Bson bson = null;
+
```
[the remark this answers is not in the document]
Comment (dwasinge) [reply src/main/java/com/redhat/labs/lodestar/model/search/StateSearchComponent.java:65]: this makes sense.  i will update to make it more readable.

**Comment by dwasinge** [comment]
This should scale ok, but it would be scaling on the mongo side since these aggregation pipelines are running on the mongo server. 

The other options were the following:

1. stream all the data and do the filtering/paging/projections/etc on the application side.
2. use a combination and have to make multiple calls to get the same data.  for example:  get all the queried data using filtering/paging/sorting and then have to make a separate call without the paging to figure out the total number of results to know the total number of pages.

**Review by mcanoy — APPROVED** [review]

**Pull request merged**